src/prommis/conn/connectivity.py

```diff
@@ -0,0 +1,581 @@
+"""
+Calculate and display model connectivity.
+
+The main thing this module does is create and process a connectivity
+matrix encoded as follows:
+
+    Units,Unit 1 name,Unit 2 Name, ..., Unit N Name
+    Arcs,,,,,,,	...,
+    Arc1 Name,	-1,	0, 0, ..., 0
+    Arc2 Name, 0 , 1, 0 ,... , 0
+
+Where each cell at the intersection of an Arc (row i) and Unit (column j)
+is either:
+  *  -1 meaning Arc(i) is an outlet of Unit(j), 
+  *  1 meaning Arc(i) is an inlet for Unit(j),
+  *  0 meaning there is no connection
+
+
+This module can be run as a script or used programmatically, using the
+public functions `create_from_matrix` and `create_from_model`.
+"""
+
+import argparse
+import csv
+from dataclasses import dataclass, field
+import enum
+import importlib
+from io import StringIO
+import logging
+from pathlib import Path
+import pprint
+import sys
+from tempfile import TemporaryFile
+from traceback import format_stack
+from typing import TextIO
+import warnings
+
+try:
+    import pyomo
+    from pyomo.network import Arc
+except ImportError as err:
+    pyomo = None
+    warnings.warn(f"Could not import pyomo: {err}")
+
+# For logging
+from prommis.conn.util.yoder import add_log_options, process_log_options
+
+# Constants
+AS_STRING = "-"
+
+# Logging
+# This variable is reassigned if run as script
+_log = logging.getLogger(__name__)
+
+
+##############
+# Classes
+##############
+
+
+class OutputFormats(enum.Enum):
+    markdown = "markdown"
+    html = "html"
+    mermaid = "mermaid"
+    csv = "csv"
+
+
+class Mermaid:
+    def __init__(self, connectivity, indent="   "):
+        self._conn = connectivity
+        self.indent = indent
+
+    def write(self, output_file: str | None, output_format: str = None):
+        if output_file is None:
+            f = StringIO()
+        else:
+            f = open(output_file, "w")
+        match output_format:
+            case OutputFormats.markdown.value:
+                f.write("# Graph\n```mermaid\n")
+                self._body(f)
+                f.write("\n```\n")
+            case OutputFormats.mermaid.value:
+                self._body(f)
+            case OutputFormats.html.value:
+                f.write(
+                    "<!doctype html>\n<html lang='en'>\n<body>\n<pre class='mermaid'>\n"
+                )
+                self._body(f)
+                f.write("</pre>\n<script type='module'>\n")
+                f.write(
+                    "import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@11/dist/mermaid.esm.min.mjs';\n"
+                )
+                f.write(
+                    "mermaid.initialize({securityLevel: 'loose', maxEdges: 2000});\n"
+                )
+                f.write("await mermaid.run();\n")
+                f.write("</script></body></html>")
+            case _:
+                raise ValueError(f"Bad output format: {output_format}")

```

Comment:
As much as it's cool to see the pattern match syntax being used, this requires Python 3.10+ and thus won't be available to us as long as we have to support Python 3.9 (about 1 year more at least).

Reply:
yeah it was fun while it lasted